controllers/awscluster_controller_unit_test.go

```diff
@@ -0,0 +1,632 @@
+			}
+			if tc.awsCluster != nil {
+				tc.awsCluster.Default()
+				g.Expect(testEnv.Create(ctx, tc.awsCluster)).To(Succeed())
+				defer func(do ...client.Object) {
+					g.Expect(testEnv.Cleanup(ctx, do...)).To(Succeed())
+				}(tc.awsCluster)
+				_, err := reconciler.Reconcile(ctx, ctrl.Request{
+					NamespacedName: client.ObjectKey{
+						Namespace: tc.awsCluster.Namespace,
+						Name:      tc.awsCluster.Name,
+					},
+				})
+				if tc.expectError {
+					g.Expect(err).ToNot(BeNil())
+				} else {
+					g.Expect(err).To(BeNil())
+				}
+			} else {
+				_, err := reconciler.Reconcile(ctx, ctrl.Request{
+					NamespacedName: client.ObjectKey{
+						Namespace: "default",
+						Name:      "test",
+					},
+				})
+				g.Expect(err).To(BeNil())
+			}
+		})
+	}
+}
+
+func TestAWSClusterReconcileOperations(t *testing.T) {
+	var (
+		reconciler AWSClusterReconciler
+		mockCtrl   *gomock.Controller
+		ec2Svc     *mock_services.MockEC2MachineInterface
+		elbSvc     *mock_services.MockELBInterface
+		networkSvc *mock_services.MockNetworkInterface
+		sgSvc      *mock_services.MockSecurityGroupInterface
+		recorder   *record.FakeRecorder
+	)
+
+	setup := func(t *testing.T, g *WithT, awsCluster *infrav1.AWSCluster) client.WithWatch {
+		t.Helper()
+		secret := &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "test-secret",
+				Namespace: "capa-system",
+			},
+			Data: map[string][]byte{
+				"AccessKeyID":     []byte("access-key-id"),
+				"SecretAccessKey": []byte("secret-access-key"),
+				"SessionToken":    []byte("session-token"),
+			},
+		}
+		csClient := fake.NewClientBuilder().WithObjects(awsCluster, secret).Build()
+
+		mockCtrl = gomock.NewController(t)
+		ec2Svc = mock_services.NewMockEC2MachineInterface(mockCtrl)
+		elbSvc = mock_services.NewMockELBInterface(mockCtrl)
+		networkSvc = mock_services.NewMockNetworkInterface(mockCtrl)
+		sgSvc = mock_services.NewMockSecurityGroupInterface(mockCtrl)
+
+		recorder = record.NewFakeRecorder(2)
+
+		reconciler = AWSClusterReconciler{
+			Client: csClient,
+			ec2ServiceFactory: func(scope.EC2Scope) services.EC2MachineInterface {
+				return ec2Svc
+			},
+			elbServiceFactory: func(elbScope scope.ELBScope) services.ELBInterface {
+				return elbSvc
+			},
+			networkServiceFactory: func(clusterScope scope.ClusterScope) services.NetworkInterface {
+				return networkSvc
+			},
+			securityGroupFactory: func(clusterScope scope.ClusterScope) services.SecurityGroupInterface {
+				return sgSvc
+			},
+			Recorder: recorder,
+		}
+		return csClient
+	}
+
+	teardown := func(t *testing.T, g *WithT) {
+		t.Helper()
+		mockCtrl.Finish()
+	}
+
+	t.Run("Reconciling an AWSCluster", func(t *testing.T) {
+		t.Run("reconcile success", func(t *testing.T) {
+			t.Run("should successfully create AWSCluster with Cluster Finalizer and LoadBalancerReady status true on AWSCluster", func(t *testing.T) {
+				g := NewWithT(t)
+				runningCluster := func(t *testing.T, g *WithT) {
```

Comment:
No need to pass `g *WithT` in arguments as you are not using it anywhere. Consider it at other places also if any